### PR TITLE
Merge native_mate into electron/electron

### PR DESCRIPTION
**Pull request by MarshallOfSound**
This PR brings in `native_mate` into the `native_mate` directory and keeps all commit history 🎉

### CODE REVIEW

**Comment by nornagon** [comment]
Is the goal to eventually replace native_mate with gin?

**Comment by deepak1556** [comment]
@nornagon yup thats the plan, but it will take some time.

**Comment by codebytere** [comment]
@deepak1556 did we figure out what that means for native promises?

File: native_mate/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+*.swp
```
Comment (nornagon) [inline native_mate/.gitignore]: we probably don't need this .gitignore

File: native_mate/LICENSE [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+The above copyright notice and this permission notice shall be
+included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
Comment (nornagon) [inline native_mate/LICENSE]: if this isn't different to Electron's license we should drop this file

File: electron.gyp [hunk 3]
```diff
@@ -10,7 +10,7 @@
   'includes': [
     'features.gypi',
     'filenames.gypi',
-    'vendor/native_mate/native_mate_files.gypi',
+    'native_mate/native_mate_files.gypi',
```
Comment (nornagon) [inline electron.gyp]: get rid of the `native_mate_files` variable in `filenames.gypi`

File: native_mate/LICENSE (hunk 2 above)
Comment (ckerr) [reply native_mate/LICENSE]: Both are MIT, but the copyright holder is different so we may want to keep it.

That's probably a question for @zcbenz to address before this gets merged :smiley:

Comment (ckerr) [reply native_mate/LICENSE]: Sublicensing from MIT to ...MIT?

Anyway, I'm not taking issue with licensing but rather that we need to give proper acknowledgment to the copyright holder, which brings us back to @zcbenz

**Comment by deepak1556** [comment]
@codebytere there is no promise helper classes from gin, the current implementation on native_mate can be easily ported to work with gin when the migration is done. That said, we should also find out if upstream has any ideas of including those helpers to gin, they might have some alternatives.

File: native_mate/LICENSE (hunk 2 above)
Comment (zcbenz) [reply native_mate/LICENSE]: Hereby I grant Electron team the permission to replace `Copyright (c) 2014 Cheng Zhao` with `Copyright (c) 2013-2018 GitHub Inc.` in the license file.

**Comment by MarshallOfSound** [comment]
Closing out, will remake

**Comment by MarshallOfSound** [comment]
Replaced by #13379

**Pull request closed without merging**